### PR TITLE
initial helpers to work with signed cookies as a session store

### PR DESCRIPTION
**Pull request by CrowdHailer**
This should solve #74 Will also be the base required for solving #76

### CODE REVIEW

File: lib/raxx/session/signed_cookie.ex [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+  defp verify_signature(payload, digest, secrets) do
+    Enum.any?(secrets, fn(secret) ->
+      secure_compare(digest, safe_digest(payload, secret))
+    end)
+  end
+
+  defp secure_compare(left, right) do
+    if byte_size(left) == byte_size(right) do
+      secure_compare(left, right, 0) == 0
```
Comment (samphilipd) [inline lib/raxx/session/signed_cookie.ex]: I'm amazed this isn't in any kind of library. Seems risky to roll your own.

File: lib/raxx/session/signed_cookie.ex [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    case Raxx.get_header(request, "cookie") do
+      nil ->
+        {:error, :no_cookies_sent}
+
+      cookie_header ->
+        case Cookie.parse(cookie_header) do
+          cookies = %{} ->
+            case Map.fetch(cookies, "#{session_config.cookie_name}") do
+              {:ok, session_cookie} ->
+                case String.split(session_cookie, "--", parts: 2) do
```
Comment (samphilipd) [inline lib/raxx/session/signed_cookie.ex]: This nesting is a bit bleurgh, can this be refactored?

File: lib/raxx/session/signed_cookie.ex [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    Enum.any?(secrets, fn(secret) ->
+      secure_compare(digest, safe_digest(payload, secret))
+    end)
+  end
+
+  defp secure_compare(left, right) do
+    if byte_size(left) == byte_size(right) do
+      secure_compare(left, right, 0) == 0
+    else
+      false
```
Comment (samphilipd) [inline lib/raxx/session/signed_cookie.ex]: You're revealing information about the length of the token here.

File: lib/raxx/session/signed_cookie.ex [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    end
+  end
+
+  defp secure_compare(<<x, left::binary>>, <<y, right::binary>>, acc) do
+    import Bitwise
```
Comment (samphilipd) [inline lib/raxx/session/signed_cookie.ex]: Does this import Bitwise multiple times?

File: lib/raxx/session/signed_cookie.ex (hunk 1 above)
Comment (CrowdHailer) [reply lib/raxx/session/signed_cookie.ex]: It's just copied from plug https://github.com/elixir-plug/plug/blob/master/lib/plug/crypto.ex#L114

Was tempted to open an issue to make crypto a core language library. but decided it was small enough to just copy paste

File: lib/raxx/session/signed_cookie.ex (hunk 3 above)
Comment (CrowdHailer) [reply lib/raxx/session/signed_cookie.ex]: yeah I thought that. but it's whats in Plugs Crypto lib

File: lib/raxx/session/signed_cookie.ex (hunk 2 above)
Comment (CrowdHailer) [reply lib/raxx/session/signed_cookie.ex]: Fair comment, though I am going to leave it for the moment. I could come up with some intermediate functions but they are not yet reused anywhere

**Pull request merged**